File: backend/src/main/java/cn/etarch/mao/harness/core/ToolMediaInjector.java
package cn.etarch.mao.harness.core;

import cn.etarch.mao.harness.llm.ChatRequest;
import cn.etarch.mao.harness.llm.LlmModelConfig;
import cn.etarch.mao.harness.tool.ImageFileSupport;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Injects synthetic user messages with image attachments after tool results,
 * following the strategy for OpenAI-compatible APIs.
 */
@Component
public class ToolMediaInjector {

    public static final String SYNTHETIC_ATTACHMENT_PROMPT = "Attached media from tool result:";

    public List<ChatRequest.Message> inject(List<ChatRequest.Message> messages,
                                            Map<String, ToolAttachment> toolAttachments,
                                            LlmModelConfig modelConfig) {
        if (messages == null || messages.isEmpty()) {
            return messages;
        }
        boolean supportsVision = modelConfig != null && Boolean.TRUE.equals(modelConfig.getSupportsVision());
        if (toolAttachments == null || toolAttachments.isEmpty()) {
            return new ArrayList<>(messages);
        }

        List<ChatRequest.Message> injected = new ArrayList<>(messages.size());
        for (ChatRequest.Message msg : messages) {
            injected.add(msg);
            if (!"tool".equals(msg.getRole()) || msg.getToolCallId() == null) {
                continue;
            }
            ToolAttachment attachment = toolAttachments.get(msg.getToolCallId());
            if (attachment == null || !ImageFileSupport.isImageMime(attachment.getMime())) {
                continue;
            }
            if (!supportsVision || attachment.getDataUri() == null || attachment.getDataUri().isBlank()) {
                continue;
            }
            injected.add(buildSyntheticUserMessage(attachment));
        }
        return injected;
    }

    private ChatRequest.Message buildSyntheticUserMessage(ToolAttachment attachment) {
        List<ChatRequest.ContentPart> parts = new ArrayList<>();
        parts.add(ChatRequest.ContentPart.builder()
                .type("text")
                .text(SYNTHETIC_ATTACHMENT_PROMPT)
                .build());
        parts.add(ChatRequest.ContentPart.builder()
                .type("image_url")
                .imageUrl(ChatRequest.ImageUrl.builder().url(attachment.getDataUri()).build())
                .build());
        return ChatRequest.Message.builder()
                .role("user")
                .content(parts)
                .build();
    }
}

File: backend-ts/src/harness/core/tool-media-injector.ts
import type { ChatMessage, LlmModelConfig } from '../llm/chat-request.js';
import type { ToolAttachment } from './tool-attachment.js';
import { ImageFileSupport } from '../tool/image-file-support.js';

export const SYNTHETIC_ATTACHMENT_PROMPT = 'Attached media from tool result:';

export class ToolMediaInjector {
  inject(
    messages: ChatMessage[] | null | undefined,
    toolAttachments: Map<string, ToolAttachment> | Record<string, ToolAttachment> | null | undefined,
    modelConfig: LlmModelConfig | null | undefined,
  ): ChatMessage[] | null | undefined {
    if (!messages || messages.length === 0) return messages;
    const supportsVision = modelConfig?.supportsVision === true;
    const map = toolAttachments instanceof Map
      ? toolAttachments
      : new Map(Object.entries(toolAttachments ?? {}));
    if (map.size === 0) return [...messages];

    const injected: ChatMessage[] = [];
    let i = 0;
    while (i < messages.length) {
      const msg = messages[i];
      if (msg.role !== 'tool') {
        injected.push(msg);
        i++;
        continue;
      }
      // 连续 tool 结果必须成组保留：网关要求 assistant.tool_calls 之后紧跟全部 tool output，
      // 中间插入 user（含工具图片）会触发 400「No tool output found for tool call」。
      const group: ChatMessage[] = [];
      while (i < messages.length && messages[i].role === 'tool') {
        group.push(messages[i]);
        i++;
      }
      injected.push(...group);
      if (!supportsVision) continue;
      for (const toolMsg of group) {
        if (!toolMsg.toolCallId) continue;
        const attachment = map.get(toolMsg.toolCallId);
        if (!attachment || !ImageFileSupport.isImageMime(attachment.mime) || !attachment.dataUri) continue;
        injected.push({
          role: 'user',
          content: [
            { type: 'text', text: SYNTHETIC_ATTACHMENT_PROMPT },
            { type: 'image_url', imageUrl: { url: attachment.dataUri } },
          ],
        });
      }
    }
    return injected;
  }
}
